compiler/codegen/codegen.cpp

```diff
@@ -953,10 +953,48 @@ static const int maxUniquifyAddedChars = 25;
 static const int maxCNameAddedChars = 20;
 static char* longCNameReplacementBuffer = NULL;
 static Map<const char*, int> uniquifyNameCounts;
-static const char* uniquifyName(const char* name,
+
+// Return the next uniquifying number for 'name'.
+// Cache the lookup result in 'elem' - speed up "call_tmp" etc.
+static int uniquifyNameNextCount(MapElem<const char*, int>*& elem,
+                                 const char* name) {
+  if (!elem) {
+    elem = uniquifyNameCounts.get_record(name);
+    if (!elem) {  // The first time we see 'name'.
+      uniquifyNameCounts.put(name, 2);
+      return 2;
+    }
+  }
+  return ++elem->value;
+}
+
+static void uniquifyName(Symbol* sym,
                                 std::set<const char*>* set1,
                                 std::set<const char*>* set2 = NULL) {
+  const char* name = sym->cname;
   const char* newName = name;
+
+  if (!sym->isRenameable()) {
+    // Record this symbol's name, in case there are renameable symbols
+    // with the same name in the future.
+
+    // If we have already seen this name before, we need to go back
+    // to that earlier-seen symbol, either renaming it or checking
+    // that its type is the same.
+    //
+    // This is currently not implemented. For now, at least detect it
+    // to avoid generating erroneous C code silently.
+    // Do this only for things local to a function, as we use
+    // multiple Symbols for the same extern at the global scope,
+    // ex c_pointer_return, qbuffer_ptr_t, QBUFFER_PTR_NULL.
+    //
+    if (set2 && (set1->find(name) != set1->end()) )
+      INT_FATAL(sym, "name conflict with a non-renameable symbol");
+
+    set1->insert(name);
+    return;
```

Comment:
To make the control flow of this function clearer, I'd recommend taking the approach of:

```c
  if (sym->isRenameable()) {
    ... do all the normal stuff ..
  } else {
    ... do the int fatal if the symbol's already in the set ...
  }
  set1->insert(name);
}
```

such that the actions that are different in each case are separated from the one that's common (inserting the name into the set).